atlasdb-api/src/main/java/com/palantir/atlasdb/persistent/api/PersistentStore.java

```diff
@@ -0,0 +1,91 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.persistent.api;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+
+import javax.annotation.Nonnull;
+
+import org.immutables.value.Value;
+
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+
+public interface PersistentStore<K, V> extends AutoCloseable {
+    @Value.Immutable
+    interface EntryFamilyHandle {
+        UUID uniqueName();
+    }
+
+    /**
+     * Gets the value associated with the entry specified by {@code key}.
+     *
+     * @param entryFamilyHandle handle to the entry family from which we want to retrieve the value
+     * @param key                entry key for which we want to retrieve the value
+     * @return the {@link Optional} containing the value or empty if there is no associated value
+     * @throws SafeIllegalArgumentException when {@code entryFamilyHandle} is a handle to a non existing entry family
+     */
+    Optional<V> get(EntryFamilyHandle entryFamilyHandle, @Nonnull K key);
+
+    /**
+     * Gets the values associated with the entries specified by {@code keys}.
+     *
+     * @param entryFamilyHandle handle to the entry family from which we want to retrieve the values
+     * @param keys               representing keys for which we want to retrieve the values
+     * @return a map from keys to values
+     */
+    Map<K, V> get(EntryFamilyHandle entryFamilyHandle, List<K> keys);
+
+    /**
+     * Stores the {@code value} for the associated {@code key} while overwriting the existing value in the specified
+     * {@code entryFamilyHandle}.
+     *
+     * @param entryFamilyHandle of the store to which we should store the entry
+     * @param key                entry key
+     * @param value              entry value
+     * @throws SafeIllegalArgumentException when {@code entryFamilyHandle} is a handle to a non existing entry family
+     */
+    void put(EntryFamilyHandle entryFamilyHandle, @Nonnull K key, @Nonnull V value);
+
+    /**
+     * Stores the entry pairs given in {@code toWrite}, overwriting the existing values.
+     *
+     * @param entryFamilyHandle of the store to which we should store the entry
+     * @param toWrite            entry pairs to write
+     * @throws SafeIllegalArgumentException when {@code entryFamilyHandle} is a handle to a non existing entry family
+     */
+    void put(EntryFamilyHandle entryFamilyHandle, Map<K, V> toWrite);
+
+    /**
+     * Creates a handle of type {@link EntryFamilyHandle} to the underlying store. Each call returns a new {@link
```

Comment:
This javadoc is somewhat roundabout and doesn't say anything that makes me understand what a handle is, and what calling `createHandle`` multiple times does internally. i.e. I can get multiple handles but are they to the same "door" so to speak? Or are they handles to different "doors". From earlier conversation it's the latter, but it's not super clear here.

The type is not relevant, you can see it in the method signature.

Maybe something like: "createSpace" where space is described as a newly initialised place to put data or something along those lines, @gmaretic what do you think?

Reply:
I implemented the `Handle` solution, if @gmaretic agrees we'll proceed with it.